### ISSUE TITLE
Support more than one report per job

### ISSUE THREAD

**Issue by tsondergaard** [jenkinsci/cobertura-plugin#82]
It would be great to be able to generate more than one coverage report and time plot. We do that now by teiggering a separate job that uses the same workspace as the first job to generate “focused” coverage report for certain areas that we are working on improving code coverage in

**Comment by jeffpearce** [comment]
Thanks for filling an issue. I'm not sure I understand what that would look like. Would there be two charts on the project and build pages? And what sorts of criteria would be used to select the addition reports?

**Comment by tsondergaard** [comment]
Yes, there would be two charts and two reports. The job would generate two coverage.xml files and then jenkins job would list them both in the cobertura configuration. That’s how I imagine it would work

**Comment by jeffpearce** [comment]
But what criteria would you use to generate the second report? Is a certain directory tree in your source?

**Comment by tsondergaard** [comment]
We generate two coverage reports - coverage_report.xml and filtered_coverage_report.xml like this:

```
$MI_SOURCE_DIR/scripts/gcovr --xml --root $MI_SOURCE_DIR --exclude=.*/tests/.* --exclude=.*/build/.* -o $MI_BUILD_DIR/coverage_report.xml $MI_SOURCE_DIR

# scripts/coverage-filter-regex is a newline separated list of regex filters
$MI_SOURCE_DIR/scripts/gcovr --xml --root $MI_SOURCE_DIR \
      --exclude=.*/tests/.* --exclude=.*/build/.* \
      --filter="$(ruby -e "puts $<.readlines.reject { |e| e =~ /^(\s*#.*|\s*)$/ }.map { |e| '.*' + e.strip }.join('|')" $MI_SOURCE_DIR/scripts/coverage-filter-regex-list)" \
      -o $MI_BUILD_DIR/filtered_coverage_report.xml $MI_SOURCE_DIR
```

The second more complex statement uses gcovr's --filter option to only include coverage for source files that match the filter. It reads a file with newline separated regexes (for readability) and joins them together with '|'  to form one big regex filter.

**Comment by macetw** [comment]
I have two different Cobertura test results from two different sets of tests. We have unit tests and mock tests. We have justification to document the coverage of both sets of tests, but at this time, this is not working.
What happens instead is that the same coverage report is shown twice.

**Comment by arcadien** [comment]
it would be great to have more than one report published for the same job. Here, javascript and C++ tests

**Comment by jsiei97** [comment]
> 
> 
> it would be great to have more than one report published for the same job. Here, javascript and C++ tests

I have a project with some python and c code that both output coverage as cobertura.xml. However the plugin only show one of them.

**Comment by jsiei97** [comment]
> 
> 
> But what criteria would you use to generate the second report? Is a certain directory tree in your source?

They have different package name in the xml file.

```
    <packages>
        <package branch-rate="0" complexity="0" line-rate="1" name="MyNiceProject">

```
So maybe show them as different "packages" on the first page?

**Comment by chauncey-garrett** [comment]
The main issue I've had with configuring multiple reports is that the first report will be overwritten by the second. 

Until multiple reports are supported, here's a solution to at least keep that from happening. The downside here is that the reports are combined, which doesn't make much sense for different apps/languages.

https://github.com/jenkinsci/cobertura-plugin/issues/92#issuecomment-367707969

**Comment by cizezsy** [comment]
@chauncey-garrett You can use the [Code Coverage API plugin](https://github.com/jenkinsci/code-coverage-api-plugin) which supports multiple reports. The latest Cobertura plugin has implemented it :)

**Comment by GeorgeSealy** [comment]
I'd definitely add my request for this feature as well. Our use case is the gradual refactoring / replacement of legacy code with a more modern, testable code base. We need to track / report on both unit test coverage overall and for the coverage for just the new code.

**Comment by gcinar** [comment]
This would be a very useful feature for repos with multiple languages

**Comment by jayahariv** [comment]
+1

**Comment by jeffpearce** [comment]
Have people requesting this tried @cizezsy's suggestion to try the Code Coverage API to see whether it meets your needs?

**Comment by jayahariv** [comment]
I am seeing the coverage breakdown by package/files with Cobertura Code Coverage API but, Code Coverage API is only showing a general high level report.

**Comment by cizezsy** [comment]
@jayahariv Code Coverage API plugin can also show the coverage breakdown by package/files. You need to click the name of the report chart to see the detail.

**Comment by jayahariv** [comment]
I click on the name of the report chart to see the details, and I was seeing below chart again. I kept  clicking on graph areas from here, but not the project name. 

Now on clicking the project name it is showing the same details as Cobertura API does.
<img width="875" alt="Screen Shot 2020-01-06 at 12 35 05 PM" src="https://user-images.githubusercontent.com/10448770/71803056-bb564c00-3085-11ea-8113-77d6cae82c43.png">

thanks @cizezsy Code Coverage API will satisfy my need.

**Comment by MarkusPalcer** [comment]
It doesn't do the job for me, as the usability of the code coverage API is near zero (which is not the scope of this issue tracker).

I'd be happy to be able to export more than one report from one Jenkins Job. Currently I will get two links to the same (the last exported) report, instead of simply two links to each of the reports.

**Comment by tsondergaard** [comment]
@cizezsy and @jayahariv - you says the "Code Coverage API" solves the problem. It is unclear to me what I need to change in my jenkinsfile. Could you help me with that? What I have today is this:

```
cobertura coberturaReportFile: 'build/**/cobertura-coverage.xml'
```

**Comment by tsondergaard** [comment]
To answer my own question, I have found out that I can also publish my coverage report with
```
publishCoverage adapters: [coberturaAdapter('cob1.xml'),coberturaAdapter('cob2.xml')], sourceFileResolver: sourceFiles('STORE_LAST_BUILD')
```

But it takes about 8 minutes to copy source files - meticulously logging both when it starts and finishes copying the source file. Also, it appears to get the reports mixed up a bit, ie where cob2.xml overlaps with cob1.xml the output shown for cob1.xml is based on the information in cob2.xml.

**Comment by jeffpearce** [comment]
I think doing it that way publishCoverage would attempt to combine the two reports (on purpose). Have you tried publishing in two separate steps? E.g.
publishCoverage adapters: [coberturaAdapter('cob1.xml')]
publishCoverage adapters: [coberturaAdapter('cob2.xml')]
That should give you two reports, and hopefully avoid the overlap

**Comment by sathnaga** [comment]
any solution to this yet? would be very helpful, thanks in advance.

> I think doing it that way publishCoverage would attempt to combine the two reports (on purpose). Have you tried publishing in two separate steps? E.g. publishCoverage adapters: [coberturaAdapter('cob1.xml')] publishCoverage adapters: [coberturaAdapter('cob2.xml')] That should give you two reports, and hopefully avoid the overlap

cobertura coberturaReportFile: 'a/coverage*.xml'
cobertura coberturaReportFile: 'b/coverage*.xml'
cobertura coberturaReportFile: 'c/coverage*.xml'
cobertura coberturaReportFile: 'd/coverage*.xml', 
cobertura coberturaReportFile: 'e/coverage*.xml'

all reports gets overwritten with last published one i.e `e`
